compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/irCodegenUtils.kt

```diff
@@ -408,23 +401,26 @@ internal fun getSignature(
    For other cases use getVisibilityAccessFlag(MemberDescriptor descriptor)
    Classes in byte code should be public or package private
 */
-fun IrClass.getVisibilityAccessFlagForClass(): Int {
-    /* Original had a check for SyntheticClassDescriptorForJava, never invoked in th IR backend. */
-    if (isOptionalAnnotationClass()) {
-        return AsmUtil.NO_FLAG_PACKAGE_PRIVATE
-    }
-    if (kind == ClassKind.ENUM_ENTRY) {
-        return AsmUtil.NO_FLAG_PACKAGE_PRIVATE
-    }
-    return if (visibility === Visibilities.PUBLIC ||
-        visibility === Visibilities.PROTECTED ||
+fun IrClass.getVisibilityAccessFlagForClass(context: JvmBackendContext): Int =
+    when {
+        isOptionalAnnotationClass() -> AsmUtil.NO_FLAG_PACKAGE_PRIVATE
+        kind === ClassKind.ENUM_ENTRY -> AsmUtil.NO_FLAG_PACKAGE_PRIVATE
+        visibility === Visibilities.PRIVATE -> AsmUtil.NO_FLAG_PACKAGE_PRIVATE
+        isClassForCallableReference -> getVisibilityAccessFlagForAnonymous(context)
+        visibility === Visibilities.PUBLIC -> Opcodes.ACC_PUBLIC
+        visibility === Visibilities.PROTECTED -> Opcodes.ACC_PUBLIC
         // TODO: should be package private, but for now Kotlin's reflection can't access members of such classes
-        visibility === Visibilities.LOCAL ||
-        visibility === Visibilities.INTERNAL
-    ) {
-        Opcodes.ACC_PUBLIC
-    } else AsmUtil.NO_FLAG_PACKAGE_PRIVATE
-}
+        visibility === Visibilities.LOCAL -> Opcodes.ACC_PUBLIC
+        visibility === Visibilities.INTERNAL -> Opcodes.ACC_PUBLIC
+        visibility === JavaVisibilities.PACKAGE_VISIBILITY -> AsmUtil.NO_FLAG_PACKAGE_PRIVATE
+        else -> error("Unexpected visibility: $visibility")
+    }
+
+private val IrClass.isClassForCallableReference: Boolean
+    get() = origin == JvmLoweredDeclarationOrigin.LAMBDA_IMPL ||
+            origin == JvmLoweredDeclarationOrigin.FUNCTION_REFERENCE_IMPL ||
+            origin == JvmLoweredDeclarationOrigin.GENERATED_SAM_IMPLEMENTATION ||
```

Comment:
Looks like we already generate `GENERATED_SAM_IMPLEMENTATION` and `GENERATED_PROPERTY_REFERENCE` classes as `PUBLIC`, so these two lines are redundant